component/azstorage/block_blob.go

```diff
@@ -653,6 +655,200 @@ func (bb *BlockBlob) WriteFromBuffer(name string, metadata map[string]string, da
+		blockList.BlockOffsetList = append(blockList.BlockOffsetList, newBlock)
+	}
+	return appendOnly, bufferSize
+}
+
+// Write : write data at given offset to a blob
+func (bb *BlockBlob) Write(name string, offset, length int64, data []byte, fileOffsets, modFileOffsets *common.BlockOffsetList) error {
+	defer log.TimeTrack(time.Now(), "BlockBlob::Write", name)
+	log.Trace("BlockBlob::Write : name %s offset %v", name, offset)
+	blobURL := bb.Container.NewBlockBlobURL(filepath.Join(bb.Config.prefixPath, name))
+	// tracks the case where our offset is great than our current file size (appending only - not modifying pre-existing data)
+	appendOnly := false
+	var dataBuffer *[]byte
+
+	// when the file offset mapping is cached we don't need to make a get block list call
+	if fileOffsets != nil && !fileOffsets.Cached {
+		var err error
+		fileOffsets, err = bb.GetFileBlockOffsets(name)
+		if err != nil {
+			return err
+		}
+	}
+
+	// case 1: file consists of no blocks (small file)
+	if fileOffsets != nil && fileOffsets.SmallFile {
+		// get all the data
+		oldData, _ := bb.ReadBuffer(name, 0, 0)
+		// update the data with the new data
+		// if we're only overwriting existing data
+		if int64(len(oldData)) >= offset+length {
+			copy(oldData[offset:], data)
+			dataBuffer = &oldData
+			// else appending and/or overwriting
+		} else {
+			// if the file is not empty then we need to combine the data
+			if len(oldData) > 0 {
+				// new data buffer with the size of old and new data
+				newDataBuffer := make([]byte, offset+length)
+				// copy the old data into it
+				// TODO: better way to do this?
+				copy(newDataBuffer, oldData)
+				oldData = nil
+				// overwrite with the new data we want to add
+				copy(newDataBuffer[offset:], data)
+				dataBuffer = &newDataBuffer
+			} else {
+				dataBuffer = &data
+			}
+		}
+		// WriteFromBuffer should be able to handle the case where now the block is too big and gets split into multiple blocks
+		err := bb.WriteFromBuffer(name, nil, *dataBuffer)
+		if err != nil {
+			log.Err("BlockBlob::Write : Failed to upload to blob %s ", name, err.Error())
+			return err
+		}
+		// case 2: given offset is within the size of the blob - and the blob consists of multiple blocks
+		// case 3: new blocks need to be added
+	} else {
+		modifiedBlockList, oldDataSize, exceedsFileBlocks := fileOffsets.FindBlocksToModify(offset, length)
+		// keeps track of how much new data will be appended to the end of the file (applicable only to case 3)
+		newBufferSize := int64(0)
+		// case 3?
+		if exceedsFileBlocks {
+			// get length of blockID in order to generate a consistent size block ID so storage does not throw
+			existingBlockId, _ := base64.StdEncoding.DecodeString(fileOffsets.BlockOffsetList[0].Id)
+			blockIdLength := len(existingBlockId)
+			appendOnly, newBufferSize = bb.createNewBlocks(modifiedBlockList, fileOffsets, offset, length, int64(blockIdLength))
```

Comment:
We are modifying existing blocks here and then below code we modify the last block seperatly before doing stage and commit. This looks little duplicate stuff as base objective is to modify some/all data in a given block. We shall think of simplifying this code further.

Reply:
CreateNewBlocks is only generating new blocks I see then where is the loop to modify one or more blocks which are getting affected due to write operation. Below code modifies only one block.

Reply:
I'm not sure I follow here - we can discuss offline

Reply:
Sure we can discuss, base idea is we are iterating these two block lists multiple times in flow so we can think of simplifying the solution.